### PR TITLE
Add deconstruction lang reference

### LINKED ISSUE #28091: Possible improvement: user-defined type deconstructors with same arity
> Multiple Deconstruct methods having the same number of parameters are ambiguous. You must be careful to define Deconstruct methods with different numbers of parameters, or "arity". Deconstruct methods with the same number of parameters cannot be distinguished during overload resolution.

It is not needed to avoid same-arity deconstructors. This part can be uplifted to point out classic method call w/o compiler ability leverage although then `Deconstruct` does not differ from any other method with `out` parameters. That since (FMPOV) it is likely that in future, it will be possible to specify overload by variable **explicit type provision**.

```csharp
class SameArityDeconstructors
. {
.     public void Deconstruct ( out int i, out int y ) { i = 3; y = 8; }
.     public void Deconstruct ( out string s, out string z ) { s = "s"; z = "z"; }
. }
> new SameArityDeconstructors ().Deconstruct ( out int i, out int y );
> i
3
> y
8
```

It can be said also if type can have for instance 1 .dcnstr with arity of 2, then other with arity of 3, it is not so fine to have suddenly another methods, let name them `NotSupportedDeconstruc`, for resting 2 .dcnsrts with arity of 4.

#### Explicit type provision does solve nothing by now
```csharp
(string s, string z) = new SameArityDeconstructors ();
(1,24): error CS0121: The call is ambiguous between the following methods or properties: 'SameArityDeconstructors.Deconstruct(out int, out int)' and 'SameArityDeconstructors.Deconstruct(out string, out string)'
```
---

Also _arity_  can be considered fine, nice word. It does not have be enclosed in quotation marks.

Edit by @BillWagner: Move to language reference.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: f3fbaed1-06d8-16f0-70d1-97792daccfd5
* Version Independent ID: 830019b6-d46c-2867-9dbd-dd89a77dd7a4
* Content: [C# Language reference](https://docs.microsoft.com/dotnet/csharp/language-reference/)
* Content Source: [https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/index.md](https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/index.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 350758](https://dev.azure.com/msft-skilling/Content/_workitems/edit/350758)

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #28091

The scenario described in the referenced issue is an advanced scenario, and didn't belong in the fundamentals section. There wasn't an article on *deconstruction expressions* in the language reference. This PR adds that. In addition, there wasn't an article in the language reference on the *discard* token. This PR adds that as well.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/operators/deconstruction.md](https://github.com/dotnet/docs/blob/184e4e1b3513a0f39ec5083889a470c481f638ae/docs/csharp/language-reference/operators/deconstruction.md) | [Deconstruction expression - Extract properties of fields from a tuple or other user-defined type](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/deconstruction?branch=pr-en-us-44023) |
| [docs/csharp/language-reference/toc.yml](https://github.com/dotnet/docs/blob/184e4e1b3513a0f39ec5083889a470c481f638ae/docs/csharp/language-reference/toc.yml) | [docs/csharp/language-reference/toc](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/toc?branch=pr-en-us-44023) |
| [docs/csharp/language-reference/tokens/discard.md](https://github.com/dotnet/docs/blob/184e4e1b3513a0f39ec5083889a470c481f638ae/docs/csharp/language-reference/tokens/discard.md) | [Discard - A `_` acts as a placeholder for a variable](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/discard?branch=pr-en-us-44023) |
| [docs/csharp/language-reference/tokens/index.md](https://github.com/dotnet/docs/blob/184e4e1b3513a0f39ec5083889a470c481f638ae/docs/csharp/language-reference/tokens/index.md) | [C# Special Characters](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/index?branch=pr-en-us-44023) |

### CODE REVIEW

**Comment by BillWagner** [comment]
ping @adegeo for review

File: docs/csharp/language-reference/tokens/index.md [hunk 1]
```diff
@@ -16,9 +16,11 @@ ms.assetid: 4c5c0539-2e37-40b7-91ce-75af5aabd3f9
 
 # C# Special Characters
 
-Special characters are predefined, contextual characters that modify the program element (a literal string, an identifier, or an attribute name) to which they are prepended. C# supports the following special characters:
+Special characters are predefined, contextual characters that modify the program element (a literal string, an identifier, or an attribute name) to which they're prepended. C# supports the following special characters:
 
 - [@](./verbatim.md), the verbatim identifier character.
 - [$](./interpolated.md), the interpolated string character.
+- ["""](./raw-string.md), A sequence of three or more `"` characters provides the delimiters for a raw string literal.
+- [_](./discard.md), a `_` represents a *discard*, a placeholder for an unused variable.
```
Comment (gewarren) [inline docs/csharp/language-reference/tokens/index.md:24]:
```suggestion
- [_](./discard.md), a `_` character represents a *discard*, which is a placeholder for an unused variable.
```

File: docs/csharp/language-reference/toc.yml [hunk 2]
```diff
@@ -322,6 +322,9 @@ items:
   - name: with expression
     href: ./operators/with-expression.md
     displayName: "records, copy"
+  - name: Deconstruction expressions
```
Comment (gewarren) [inline docs/csharp/language-reference/toc.yml:325]:
```suggestion
  - name: Deconstruction expression
```

File: docs/csharp/language-reference/operators/deconstruction.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
+---
+title: "Deconstruction expression - extract properties or fields from a tuple or other type"
+description: "Learn about deconstruction expressions: expressions that extract individual properties or fields from a tuple or user defined type into discrete expressions."
+ms.date: 12/17/2024
+---
+# Deconstruction expression - Extract properties of fields from a tuple or other user defined type
```
Comment (gewarren) [inline docs/csharp/language-reference/operators/deconstruction.md:6]:
```suggestion
# Deconstruction expression - Extract properties of fields from a tuple or other user-defined type
```

File: docs/csharp/language-reference/operators/deconstruction.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+A *deconstruction expression* extracts data fields from an instance of an object. Each discrete data element is written to a distinct variable, as shown in the following example:
+
+:::code language="csharp" source="./snippets/shared/Deconstruction.cs" id="TupleDeconstruction":::
+
+The preceding code snippet creates a [tuple](../builtin-types/value-tuples.md) that has two integer values, `X` and `Y`. The second statement *deconstructs* that tuple and stores the tuple elements in discrete variables `x`, and `y`.
```
Comment (gewarren) [inline docs/csharp/language-reference/operators/deconstruction.md:12]:
```suggestion
The preceding code snippet creates a [tuple](../builtin-types/value-tuples.md) that has two integer values, `X` and `Y`. The second statement *deconstructs* that tuple and stores the tuple elements in discrete variables `x` and `y`.
```

File: docs/csharp/language-reference/operators/deconstruction.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+:::code language="csharp" source="./snippets/shared/Deconstruction.cs" id="TupleDeconstructionWithDiscard":::
+
+In the preceding example, the `Y` and `label` members are discarded. You can specify multiple discards in the same deconstruction expression. You can use discards for all the members of the tuple. The following example is legal, although not useful:
+
+:::code language="csharp" source="./snippets/shared/Deconstruction.cs" id="AllDiscards":::
+
+## Record deconstruction
+
+[Record](../builtin-types/record.md) types that have a [primary constructor](../builtin-types/record.md#positional-syntax-for-property-definition) support deconstruction for positional parameters. The compiler synthesizes a `Deconstruct` method that extracts the properties synthesized from positional parameters in the primary constructor. The compiler synthesized `Deconstruction` method doesn't extract properties declared as properties in the record type.
```
Comment (gewarren) [inline docs/csharp/language-reference/operators/deconstruction.md:26]:
```suggestion
[Record](../builtin-types/record.md) types that have a [primary constructor](../builtin-types/record.md#positional-syntax-for-property-definition) support deconstruction for positional parameters. The compiler synthesizes a `Deconstruct` method that extracts the properties synthesized from positional parameters in the primary constructor. The compiler-synthesized `Deconstruction` method doesn't extract properties declared as properties in the record type.
```

File: docs/csharp/language-reference/operators/deconstruction.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+The `record` shown in the following code declares two positional properties, `SquareFeet` and `Address`, along with another property, `RealtorNotes`:
+
+:::code language="csharp" source="./snippets/shared/Deconstruction.cs" id="RecordDeconstruction":::
+
+When you deconstruct a `House` object, all positional properties, and only positional properties, are deconstructed, as shown in the following example:
+
+:::code language="csharp" source="./snippets/shared/Deconstruction.cs" id="RecordDeconstructionUsage":::
+
+You can make use of this behavior to specify which properties of your record types are part of the compiler synthesized `Deconstruct` method.
```
Comment (gewarren) [inline docs/csharp/language-reference/operators/deconstruction.md:36]:
```suggestion
You can make use of this behavior to specify which properties of your record types are part of the compiler-synthesized `Deconstruct` method.
```

File: docs/csharp/language-reference/operators/deconstruction.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+The following code declares a `Point3D` struct that has two `Deconstruct` methods:
+
+:::code language="csharp" source="./snippets/shared/Deconstruction.cs" id="StructDeconstruction":::
+
+The first method supports deconstruction expressions that extract all three axes values: `X`, `Y`, and `Z`. The second method supports deconstructing only the planar values: `X` and `Y`. The first method has an *arity* of 3; the second has an arity of 2.
```
Comment (gewarren) [inline docs/csharp/language-reference/operators/deconstruction.md:46]:
```suggestion
The first method supports deconstruction expressions that extract all three axis values: `X`, `Y`, and `Z`. The second method supports deconstructing only the planar values: `X` and `Y`. The first method has an *arity* of 3; the second has an arity of 2.
```

File: docs/csharp/language-reference/operators/deconstruction.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+The preceding section described the compiler synthesized `Deconstruct` method for `record` types with a primary constructor. You can declare more `Deconstruct` methods in record types. These methods can either add other properties, remove some of the default properties, or both. You can also declare a `Deconstruct` that matches the compiler synthesized signature. If you declare such a `Deconstruct` method, the compiler doesn't synthesize one.
```
Comment (gewarren) [inline docs/csharp/language-reference/operators/deconstruction.md:48]:
```suggestion
The preceding section described the compiler-synthesized `Deconstruct` method for `record` types with a primary constructor. You can declare more `Deconstruct` methods in record types. These methods can either add other properties, remove some of the default properties, or both. You can also declare a `Deconstruct` that matches the compiler-synthesized signature. If you declare such a `Deconstruct` method, the compiler doesn't synthesize one.
```

File: docs/csharp/language-reference/operators/deconstruction.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+Typically, multiple `Deconstruct` methods for the same type have different numbers of parameters. You can create multiple `Deconstruct` methods providing the signatures have different parameter types. As long as the compiler can determine one unique `Deconstruct` method for a deconstruction expression, the multiple `Deconstruct` methods are allowed. However, in many cases, too many `Deconstruct` methods can lead to ambiguity errors and misleading results.
```
Comment (gewarren) [inline docs/csharp/language-reference/operators/deconstruction.md:50]:
```suggestion
Multiple `Deconstruct` methods are allowed as long as the compiler can determine one unique `Deconstruct` method for a deconstruction expression. Typically, multiple `Deconstruct` methods for the same type have different numbers of parameters. You can also create multiple `Deconstruct` methods that differ by parameter types. However, in many cases, too many `Deconstruct` methods can lead to ambiguity errors and misleading results.
```

Comment (gewarren) [reply docs/csharp/language-reference/operators/deconstruction.md:50]: Restructured it a bit. Take it or leave it.

File: docs/csharp/language-reference/operators/deconstruction.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+## C# language specification
+
+For more information, see the following sections of the [C# Standard](~/_csharpstandard/standard/expressions.md#127-deconstruction):
```
Comment (gewarren) [inline docs/csharp/language-reference/operators/deconstruction.md:54]:
```suggestion
For more information, see the deconstruction section of the [C# Standard](~/_csharpstandard/standard/expressions.md#127-deconstruction).
```

**Review by gewarren — APPROVED** [review]

**Pull request merged**